File: Parcial/Carrito_C#/CarritoClient/Form1.cs
using CarritoServices.Models;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Windows.Forms.VisualStyles;

namespace CarritoClient
{
    public partial class Form1 : Form
    {

        // Nuestras dependencias, en este caso solo tenemos dtArticulos que es de tipo DataTable
        #region Dependencias
        private DataTable dtArticulos;
        #endregion

        // En esta region vamos a Inicializar la DataTable
        #region Incializar
     
        public void InicializarDataTable()
        {
            // Creamos un objeto de tipo DataTable
            DataTable dt = new DataTable();

            // Le creamos las columnas
            dt.Columns.Add("Nombre", typeof(string));
            dt.Columns.Add("Precio", typeof(double));

            // La DataTable que creamos se la pasamos por parametro a la funcion setDataTableArticulos
            this.SetDataTableArticulos(dt);
        }
        #endregion

        // Simplemente es nuestro Constructor
        #region Constructor
        public Form1()
        {
            InitializeComponent();
        }

        #endregion

        // Nuestras acciones
        #region Acciones

        // Esta funcion va a apuntar nuestra DataTable a el DataGrid que tengamos en nuestra app de Windows Forms
        public void SetDataTableArticulos(DataTable dt)
        {
            this.dtArticulos = dt;
            data_grid_Articulos.DataSource = dt;
        }

        // Esta funcion hace las operaciones del Carrito
        public void OperacionesCarrito()
        {
            // Variable para sacar el total de los Productos
            double totalProductos = 0;

            // Variable para sacar el total de los Productos que empiecen con R
            double totalProductosR = 0;

            // Variable para sacar la cantidad de Productos que emipiecen con R
            double contadorProductosR = 0;

            // Recorremos nuestra DataTable
            foreach (DataRow renglon in this.dtArticulos.Rows)
            {
                // Sumamos el total de los productos que se vayan agregando
                totalProductos += Convert.ToDouble(renglon["Precio"]);

                // Obtenemos el nombre de cada producto de nuestro Row (renglon)
                string nombre = renglon["Nombre"].ToString();

                // Si el nombre de nuestro renglon empieza con R hacemos lo siguiente
                if (nombre.ToUpper().StartsWith("R"))
                {
                    // Le sumamos a la variable totalProductosR  
                    totalProductosR += Convert.ToDouble(renglon["Precio"]);

                    // Incrementamos en 1 el contador de Productos que empiecen con R
                    contadorProductosR++;
                }
            }

            // Sacamos los promedios
            double promedio = totalProductos / this.dtArticulos.Rows.Count; // Como es sobre el total, al hacer Rows.Count contabilizamos la cantidad total de FILAS de nuestro DataTable
            double promedioR = totalProductosR / contadorProductosR;

            // Visualizamos los resultados de la suma de Totales
            lbl_Total.Text = totalProductos.ToString("#0.00");
            lbl_totalR.Text = totalProductosR.ToString("#0.00");

            // Visualizamos los resutlados de los promedios
            lbl_Promedio.Text = promedio.ToString("#0.00");

            // Para el promedio de los productos con R hacemos una condicion para evitar la division por cero 0
            if (contadorProductosR > 0) {

                // Si la cantidad es mayor a cero 0 que me muestre el promedio
                lbl_PromedioR.Text = promedioR.ToString("#0.00");
            } else
            {
                // Si es cero 0 que me muestre hardcodeado "0.00"
                lbl_PromedioR.Text = "0.00";
            }


        }

        // Esta funcion sirve para borrar lo que se escribio en los TextBoxes
        public void borrarTextBoxes()
        {
            txt_descripcion.Text = "";
            txt_precio.Text = "";
        }
        #endregion

        // Region para los EVENTOS de cada componente
        #region Eventos

        // Evento click para guardar el producto, sumar los totales y promedios
        private void btn_GuardarProducto_Click(object sender, EventArgs e)
        {
            // Tomo el valor del TextBox de descripcion y lo guardo en una variable
            string nombre = txt_descripcion.Text;

            // Valido que se haya ingresado un numero en el TextBox de precio
            try
            {
                // Tomo el valor del TextBox de precio 
                string precioS = txt_precio.Text;

                // Lo parseo a Double
                double precio = Double.Parse(precioS);

                // Creo el nuevo articulo con los valores dados
                Articulo nuevoArticulo = new Articulo(nombre, precio);

                // Agrego una FILA nueva a la DataTable, dtArticulos, con los valores del nuevo Articulo
                this.dtArticulos.Rows.Add(nuevoArticulo.description, nuevoArticulo.price);

                // Realizo las operaciones correspondientes
                this.OperacionesCarrito();

                // Finalmente, reseteo los TextBox
                this.borrarTextBoxes();
            }
            catch (Exception)
            {
                // Si en el TextBox de precio se puso un numero que me muestre un mensaje de Error
                MessageBox.Show("El precio debe ser un numero");
            }
        }

        // Evento load para inicializar la DataTable 
        /*
          Llamo a mi funcion IncializarDataTable que hace lo siguiente:
            - Crea la DataTable
            - Le define las COLUMNAS a la DataTable
            - Llama a la funcion SetDataTableArticulos que hace lo siguiente: 
                 - Le pasamos nuestra DataTableCreada
                 - La DataTable que creamos como DEPENDENCIA tendra el valor de nuestra DataTable creada y la pasamos por parametro
                 - A la DataGrid le asignamos nuestra DataTable creada en nuestro DataSource
                 - De esta manera nuestra DataTable que creamos como DEPENDENCIA, dtArticulos, tiene una DataTable con una estrectura de COLUMNAS y apunta a un DataGridView
         */
        private void Form1_Load(object sender, EventArgs e)
        {
            this.InicializarDataTable();
        }
        #endregion



    }
}

File: Parcial/Carrito_VB/Carrito/Form1.vb
Public Class Form1

    'En esta region estaran las dependencias necesarios para poder trabajar'
#Region "Dependencias"

    'Nuestra DataTable para que funcione el Programa'
    Private dtProductos As DataTable
#End Region

    'En esta region estan todas las funcioes para que funcione nuestra aplicacion de windows Forms' 
#Region "Acciones"

    'Esta funcion lo que hace es que la DataTable de Articulos APUNTE a nuestra DataGrid'
    Private Sub SetDataTableArticulos(dt As DataTable)

        'Nuestra DataTable (dtProductos) que creamos como dependencia va a recibir un objeto de tipo DataTable'
        Me.dtProductos = dt

        'A su vez nuestro componente DataGrid_Articulos va a recibir como DataSource nuestra DataTable de dependencia'
        dataGrid_Articulos.DataSource = dt
    End Sub

    'Esta funcion inicializa la DataTable con las COLUMNAS' 
    Private Sub InicializarDataTable()

        'Creamos el objeto de tipo DataTable' 
        Dim dt As New DataTable

        'Le definimos a nuestra DataTable las COLUMNAS'
        dt.Columns.Add("Nombre")
        dt.Columns.Add("Precio")

        'A nuestra DataTable creada la seteamos con el DataGrid'
        Me.SetDataTableArticulos(dt)
    End Sub

    'Esta funcion me suma los productos'
    Private Sub OperacionesCarrito()

        'Variables para sumar el total de productos'
        Dim totalProductos As Double = 0

        'Variables para sumar el total de productos con la letra R'
        Dim totalProductosR As Double = 0

        'Variable contadora de productos con la letra R'
        Dim contadorProductosR = 0

        'Recorro mi DataTable de Productos (Por cada RENGLON, que es de tipo DataRow, de mis Filas de la DataTable de dtProductos HACE lo siguiente)' 
        For Each renglon As DataRow In Me.dtProductos.Rows

            'Hago la suma de total de los productos'
            totalProductos += renglon("Precio")

            'Obtengo el nombre de cada Renglon'
            Dim nombre As String = renglon("Nombre")

            'Si el nombre obtenido de renglon empieza con R (le pongo ToUpper para que siempre sea con Mayuscula y evitar el case sensitive) hago lo siguiente'
            If nombre.ToUpper().StartsWith("R") Then

                'Sumo los importes de los productos que empiecen con R en mi variable totalProductos'
                totalProductosR += renglon("Precio")

                'Incremento la cantidad de Productos con la letra R a 1'
                contadorProductosR += 1
            End If
        Next

        'Saco los promedios'
        Dim promedio = totalProductos / Me.dtProductos.Rows.Count 'Como es del total de productos de mi DataTable con usar Rows.Count se cuentan todas las columnas'
        Dim promedioR = totalProductosR / contadorProductosR

        'Visualizo los resultados de los acumuladores'
        lbl_Total.Text = totalProductos.ToString("#0.00")
        lbl_totalR.Text = totalProductosR.ToString("#0.00")

        'Visualizo los resultados de los promedios'
        lbl_Promedio.Text = promedio.ToString("#0.00")

        'En el caso de ProductosR para evitar la division con 0 creo un condicional'
        If contadorProductosR > 0 Then

            'Si el contador de productos con R es mayor a 0 entonces ahi muestro el promedio'
            lbl_PromedioR.Text = promedioR.ToString("#0.00")
        Else
            'Si es cero que solo muestre un 0.00'
            lbl_PromedioR.Text = "0.00"
        End If



    End Sub


#End Region

    'Esta region son para los eventos de nuestros componentes que van a funcionar con las funciones de la region anterior'
#Region "Eventos"

    'Cuando el Form se cargue, ahi inicializo la DataTable' 
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        ' Llamo a mi funcion IncializarDataTable que hace lo siguiente:
        ' - Crea la DataTable
        ' - Le define las COLUMNAS a la DataTable
        ' - Llama a la funcion SetDataTableArticulos que hace lo siguiente: 
        '     - Le pasamos nuestra DataTableCreada
        '     - Le DataTable que creamos como DEPENDENCIA tendra el valor de nuestra DataTableCreada y que le pasamos por parametro
        '     - A la DataGrid le asignamos nuestra DataTable creada en nuestro DataSource
        '     - De esta manera nuestra DataTable que creamos como DEPENDENCIA, dtProductos, tiene una DataTable con una estrectura de COLUMNAS y apunta a un DataGridView
        Me.InicializarDataTable()
    End Sub

    Private Sub btn_Agregar_Producto_Click(sender As Object, e As EventArgs) Handles btn_Agregar_Producto.Click

        'Tomamos el nombre y el precio de nuestros TextBoxes'
        Dim nombre = txt_Nombre.Text
        Dim precio = Double.Parse(txt_Precio.Text)

        'Agregamos una nueva FILA (Rows) a nuestra DataTable dtProductos con los valores obtenidos'
        Me.dtProductos.Rows.Add(nombre, precio)

        'Llamos a la funcion OperacionesCarrito, de la region ACCIONES, que hace las operaciones internas del Programa.'
        Me.OperacionesCarrito()
    End Sub


#End Region


End Class
